### PR TITLE
Resolves #3682: leaf node fails to reconnect, due to ping messages being held off indefinitely

### LINKED ISSUE #3682: Leaf node fails to reconnect, due to ping messages being held off indefinitely
## Defect

We are using a LeafNode NATS server to connect to a cluster, and see a strange effect which prevents the LeafNode reconnecting properly in the event that its link to the cluster goes down.

The LeafNode has two local clients which connect to it, and which generate traffic on a continuous basis. And in short, this continuous client traffic results in the outbound Pings from LeafNode to cluster being held off/delayed indefinitely, with the message "Leaf Node Ping Timer", "Delaying Ping due to client activity". And because the Pings are held off, the LeafNode doesn't detect a stale connection, and hence doesn't close the connection and attempt to reconnect.

I believe I understand the issue.

In NATS server client.go:: processPingTimer() there is a check to test whether to delay sending an outgoing ping, in two cases:

- we recently (within specified pingInterval) received a data message or sub/unsub message from the remote end (Client activity)
- we recently received a ping from the remote end (Remote ping)

This makes perfect sense: incoming receive messages mean we still have a link and don't need to send a ping.

However, the first test above is derived from the client.last field ("last packet" time) which is set in two cases:

- when the readLoop parsing determines we have received a new message or sub/unsub
- in flushClients() routine, which is called when we have received some messages that need to be forwarded to other client connections

But I believe that this second case is incorrect: we should only hold off pings when there has been receive traffic on the specified connection, and this isn't so in the second case. We received traffic on one connection, but are resetting the c.last field on another connection where we are forwarding/sending the message.

If I remove the line of code in flushClients() (about line 1113) that updates cp.last, then the Stale Client Connection fires fine in my testing.

#### Versions of `nats-server` and affected client libraries used:
Latest master NATS server as of 2/12/2022, eg git commit c4c87612

#### OS/Container environment:
Linux ARM64 - but not a factor

#### Steps or code to reproduce the issue:
Configure leaf node server with local clients sending traffic that an external client is subscribed for, and then break the connection to the external NATS cluster

#### Expected result:
Leaf node server should automatically trigger a reconnect when the connection to an external NATS cluster is lost, with the detection time as configured by ping_interval and ping_max parameters.

#### Actual result:
Leaf node server does not automatically reconnect. Instead, it remains in a zombie state indefinitely. 
To be more precise, it might possibly recover when the network stack TCP keepalive timeout expires (which by default is 2 hours), but that is much too long to be useful

### PR DESCRIPTION
**Pull request by sandykellagher**
This contribution is my original work and I license the work to the project under the [Apache 2 license](https://github.com/nats-io/nats-server/blob/main/LICENSE)

### Changes proposed in this pull request:

 - don't set cp.last in client::flushClients()
 -
 -

/cc @nats-io/core

### CODE REVIEW

**Comment by derekcollison** [comment]
The last activity serves a role here. Agree we should be sending pings in the issue posted but not sure this is correct fix. See broken tests.

**Comment by sandykellagher** [comment]
processPingTimer already ensures that pings are not delayed/held off for Gateway connections. Should it do the same for Leaf Node connections? That would be a simple change, provided it doesn't have other implications...

**Comment by sandykellagher** [comment]
So I guess ConnInfo.LastActivity means either transmit or receive message. Hence the update of cp.last in flushClients().

Would it be worth having separate LastRxActivity and LastTxActivity fields?

**Comment by derekcollison** [comment]
Yes I believe we should consider that for soliciting leafnodes that they always ping to make sure they trip a write to detect a dead socket.

**Comment by derekcollison** [comment]
Let me know if you would want to take a stab at this, but I would be happy to take a look.

File: server/client.go [hunk 1]
```diff
@@ -4557,8 +4558,8 @@ func (c *client) processPingTimer() {
 	now := time.Now()
 	needRTT := c.rtt == 0 || now.Sub(c.rttStart) > DEFAULT_RTT_MEASUREMENT_INTERVAL
 
-	// Do not delay PINGs for GATEWAY connections.
-	if c.kind != GATEWAY {
+	// Do not delay PINGs for GATEWAY or LEAF connections.
+	if c.kind != GATEWAY && c.kind != LEAF {
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/client.go:4562]:
```suggestion
	if c.kind != GATEWAY && !c.isSpokeLeafNode() { ...
```

**Review by derekcollison — CHANGES_REQUESTED** [review]
We are close.

**Comment by derekcollison** [comment]
If easier just abandon this one and create new fresh from main with the two line change we agreed upon.

**Pull request closed without merging**